[CCIP Merge] Capabilities fix [CCIP-2943]

asoliman92 opened the pull request:
Fixing ccip capabilities to be merged into chainlink

cedric-cordenier committed:
[WIP] Fix compilation for chainlink.go

asoliman92 committed:
[WIP] Fix compilation for launcher, diff

Make application.go ready for adding more fixes

asoliman92 committed:
Fix launcher tests

makramkd commented on core/capabilities/ccip/launcher/diff.go:
```diff
@@ -91,19 +87,14 @@ func compareDONs(
-	state registrysyncer.State,
-) (map[registrysyncer.DonID]kcr.CapabilitiesRegistryDONInfo, error) {
-	ccipDONs := make(map[registrysyncer.DonID]kcr.CapabilitiesRegistryDONInfo)
+	ccipCapability registrysyncer.Capability, // This will change to registrysyncer.Capability
```
Can remove comment?

makramkd commented on core/capabilities/ccip/launcher/diff.go:
```diff
@@ -113,28 +104,22 @@ func filterCCIPDONs(
 // checkCapabilityPresence checks if the capability with the given version and
 // labelled name is present in the given capability registry state.
```
```suggestion
// checkCapabilityPresence checks if the capability with the given capability ID
// is present in the given capability registry state.
```

makramkd commented on core/capabilities/ccip/launcher/diff.go:
```diff
@@ -113,28 +104,22 @@ func filterCCIPDONs(
-	capabilityLabelledName string,
-	state registrysyncer.State,
-) (kcr.CapabilitiesRegistryCapabilityInfo, error) {
+	capabilityID string,
```
Feels less error prone to make this structured ID an actual structure rather than just a `string`. 

@cedric-cordenier wdyt? Maybe something like:

```golang
type CapabilityID struct {
    Name string
    Version string
}

func (cid CapabilityID) String() string {
    // perhaps some semver regex validation using the semver package
    return fmt.Sprintf("%s@%s", cid.Name, cid.Version)
}
```

asoliman92 replied:
I said a similar comment on call with @cedric-cordenier last week as well :)

asoliman92 replied:
This should not be a blocker for this PR though.

cedric-cordenier replied:
Yes, I'm not against this 👍

makramkd replied:
@asoliman92 can you create a ticket in the backlog for this if you're not going to do it in this PR

asoliman92 replied:
Done

makramkd commented on core/services/chainlink/application.go:
```diff
@@ -231,7 +272,14 @@ func NewApplication(opts ApplicationOpts) (Application, error) {
 
 		registrySyncer, err := registrysyncer.New(
 			globalLogger,
-			externalPeerWrapper,
+			func() (p2ptypes.PeerID, error) {
+				p := externalPeerWrapper.GetPeer()
```
Why are we creating two registry syncers?

makramkd replied:
I'm guessing it has something to do with new functionality, but seems confusing. Can we add comments clearly explaining the need for these two?

asoliman92 replied:
CC @cedric-cordenier

cedric-cordenier replied:
It comes down to the fact that the registrysyncer needs a peerID, and the peerID it needs is different depending on whether we're servicing CCIP or Keystone (in the Keystone case it's the external peer, for CCIP it's not necessary).

@makramkd Is this always going to be the case, or do you see a situation where the syncer might need to know what node it is (that's what the peerID is used for)?

That being said, we should only have one registrysyncer running at any given time.

makramkd replied:
@cedric-cordenier CCIP doesn't need all the other capabilities services right now; we just need a registry syncer. Is there a way to decouple the registry syncer from the stuff that needs a peerID to function?

ettec committed:
ks-409 fix the mock trigger to ensure events are sent (#14047)

asoliman92 committed:
WIP fix ocr3 tests

makramkd committed:
core/capabilities/ccip: use OCR offchain config (#1264)

We want to define and use the appropriate OCR offchain config for each
plugin.

Requires https://github.com/smartcontractkit/chainlink-ccip/pull/36/

asoliman92 committed:
Add ccip to job orm

asoliman92 committed:
Add capabilities directory under BUSL license

cedric-cordenier committed:
Prep to instantiate separate registrysyncer for CCIP

cedric-cordenier committed:
[chore] Change registrysyncer config to bytes

asoliman92 committed:
Move registrySyncer creation into ccip delegate

asoliman92 committed:
Fix launcher diff tests after changing structs in syncer

asoliman92 committed:
Fix linting

asoliman92 committed:
MAke simulated backend client work with chains other than 1337

makramkd commented on core/capabilities/ccip/delegate.go:
```diff
@@ -99,6 +106,25 @@ func (d *Delegate) ServicesForSpec(ctx context.Context, spec job.Job) (services
 
+	cfg := d.capabilityConfig
+	rid := cfg.ExternalRegistry().RelayID()
+	registryAddress := cfg.ExternalRegistry().Address()
```
Nit: you can inline this

makramkd commented on core/capabilities/ccip/launcher/diff_test.go:
```diff
@@ -23,12 +24,39 @@ const (
 	ccipCapName       = "ccip"
 )
 
+var defaultCapability = getCapability(ccipCapName, ccipCapVersion)
+var newCapability = getCapability(ccipCapName, ccipCapNewVersion)
```
Merge these var decls into one `var` decl

makramkd commented on core/capabilities/ccip/launcher/diff_test.go:
```diff
@@ -23,12 +24,39 @@ const (
+	return ragep2ptypes.PeerID(p2pkey.MustNewV2XXXTestingOnly(big.NewInt(int64(id))).PeerID())
+}
+
+var p2pID1 = getP2PID(1)
+var p2pID2 = getP2PID(2)
+
+var defaultCapCfgs = map[string]registrysyncer.CapabilityConfiguration{
+	defaultCapability.ID: registrysyncer.CapabilityConfiguration{},
+}
+var defaultRegistryDon = registrysyncer.DON{
+	DON:                      getDON(1, []ragep2ptypes.PeerID{p2pID1}, 0),
+	CapabilityConfigurations: defaultCapCfgs,
+}
```
Merge into one var decl at the top of the file for better organization

makramkd commented on core/capabilities/ccip/launcher/launcher.go:
```diff
@@ -7,20 +7,21 @@ import (
 
 	ragep2ptypes "github.com/smartcontractkit/libocr/ragep2p/types"
 
+	ccipreader "github.com/smartcontractkit/chainlink-ccip/pkg/reader"
 	ccipreaderpkg "github.com/smartcontractkit/chainlink-ccip/pkg/reader"
```
Duplicate import

makramkd commented on core/capabilities/launcher.go:
```diff
@@ -46,6 +50,42 @@ type launcher struct {
 	subServices []services.Service
 }
 
+func unmarshalCapabilityConfig(data []byte) (capabilities.CapabilityConfiguration, error) {
```
Which capability config is this unmarshalling? Not CCIP right?

asoliman92 replied:
Not CCIP AFAIU. @cedric-cordenier right?

cedric-cordenier replied:
Yeah this is only unmarshaling the Keystone capability configs

makramkd commented on core/services/chainlink/application.go:
```diff
@@ -17,6 +17,8 @@ import (
 	"go.uber.org/multierr"
 	"go.uber.org/zap/zapcore"
 
+	"github.com/smartcontractkit/chainlink/v2/core/capabilities/ccip"
```
Nit: group this with other chainlink/v2/ imports?

asoliman92 committed:
Cleaning up

asoliman92 committed:
Add capabilities types to mockery
